KAFKA-7266: Fix MetricsTest.testMetrics flakiness by increasing batch…

stanislavkozlovski opened the pull request:
The test `kafka.api.MetricsTest.testMetrics` has been failing intermittently in kafka builds.
The particular failure is in the `MessageConversionsTimeMs` metric assertion:
```
java.lang.AssertionError: Message conversion time not recorded 0.0
```

The test would intermittently fail because the conversion time took less than 1 millisecond. Such conversion time is reported as zero by the metric.
There has been work done previously (https://github.com/apache/kafka/pull/4681) to combat the flakiness of the test and while it has improved it, the test still fails sometimes.

Locally, this test failed 5 times out of 25.

### Changes
Increase the record size and use compression - both will slow down message conversion enough to have it be above 1ms. Locally this test has not failed in 200 runs and counting

stanislavkozlovski commented:
cc @ijuma @rajinisivaram for review please

rajinisivaram commented on core/src/test/scala/integration/kafka/api/IntegrationTestHarness.scala:
```diff
@@ -40,6 +40,7 @@ abstract class IntegrationTestHarness extends KafkaServerTestHarness {
   val consumerCount: Int
   val serverCount: Int
   var logDirCount: Int = 1
+  val producerBufferSize: Long = 1024L * 1024L
```
Rather than add a new field and treat as a special case, we could just get the value from `producerConfig` and use that in `createProducer`?

rajinisivaram commented on core/src/test/scala/integration/kafka/api/MetricsTest.scala:
```diff
@@ -200,11 +201,15 @@ class MetricsTest extends IntegrationTestHarness with SaslSetup {
-        sendRecords(producers.head, 1000 * iteration, 100, tp)
+        // don't increase iteration if we go over buffer limit
+        val nextBatchSize = (1000 * (iteration + 1)) * retryRecordSize
+        if (nextBatchSize < producerBufferSize)
```
Do we need this check at all? We will just be sending more batches without the check and that should be ok?

rajinisivaram reviewed (commented):
@stanislavkozlovski Thanks for the PR. Looks good, left a couple of minor comments.

ijuma commented:
Maybe set compression type to gzip to make conversion slower. That may be all that is needed.

ijuma commented on core/src/test/scala/integration/kafka/api/MetricsTest.scala:
```diff
@@ -44,7 +45,7 @@ class MetricsTest extends IntegrationTestHarness with SaslSetup {
   this.serverConfig.setProperty(KafkaConfig.AutoCreateTopicsEnableDoc, "false")
   this.producerConfig.setProperty(ProducerConfig.LINGER_MS_CONFIG, "10")
-  this.producerConfig.setProperty(ProducerConfig.BATCH_SIZE_CONFIG, "1000000")
+  this.producerConfig.setProperty(ProducerConfig.BATCH_SIZE_CONFIG, producerBufferSize.toString)
```
A way to ensure your batch is large is to use large records. Then you don't need this at all.

stanislavkozlovski replied:
True, I wrongly thought this was the minimum batch size. Should have read the documentation

stanislavkozlovski committed:
KAKFA-7266: Increase record size and compress records in metricsTest

This avoids the flakiness of the test by having the broker take more time to decompress and convert, ensuring it takes more than 1 millisecond and is recorded by the metric

stanislavkozlovski committed:
KAFKA-7266: Add comment explaining gzip config in MetricsTest

ijuma reviewed (approved):
LGTM if the test is now stable. Will leave it to @rajinisivaram to merge.

rajinisivaram commented:
retest this please

rajinisivaram reviewed (approved):
@stanislavkozlovski Thanks for the PR, LGTM. Merging to trunk.
